### ISSUE TITLE
PVC usage metrics incorrect in /stats/summary kubelet endpoint for Windows Azure Files

### ISSUE THREAD

**Issue by gracewehner** [kubernetes/kubernetes#110261]
### What happened?

Curling `https://<Node-IP>:10250/stats/summary` for a Windows node returns the wrong usage and capacity bytes for a `azurefile-csi` PVC used by a Windows pod on that node:
```
 "pods": [
  {
   "podRef": {
    "name": "win-webserver-5d5d4966f5-zrcf6",
    "namespace": "default",
    "uid": "4b2a7fec-26d4-482c-adc8-2c25baa1d054"
   },
  ...
   "volume": [
    {
     "time": "2022-05-26T23:37:24Z",
     "availableBytes": 91617021952,
     "capacityBytes": 136912564224,
     "usedBytes": 45295542272,
     "inodesFree": 0,
     "inodes": 0,
     "inodesUsed": 0,
     "name": "volume",
     "pvcRef": {
      "name": "azure-file-win",
      "namespace": "default"
     }
    }
  },

```
This looks like it is the usage of the disk on the node or something else. `azurefile-csi` PVCs used by Linux pods report the correct usage.

### What did you expect to happen?

The PVC is empty so the usedBytes should be 0 and the capacity should be 2Gi.

### How can we reproduce it (as minimally and precisely as possible)?

Apply the following to create an azurefile PVC and a windows pod that uses it:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: azure-file-win
spec:
  accessModes:
  - ReadWriteMany
  storageClassName: azurefile
  resources:
    requests:
      storage: 2Gi
---
apiVersion: v1
kind: Service
metadata:
  name: win-webserver
  labels:
    app: win-webserver
spec:
  ports:
    # the port that this service should serve on
  - port: 80
    targetPort: 80
  selector:
    app: win-webserver
  type: LoadBalancer
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: win-webserver
  name: win-webserver
spec:
  selector:
    matchLabels:
      app: win-webserver
  replicas: 1
  template:
    metadata:
      labels:
        app: win-webserver
      name: win-webserver
    spec:
      containers:
      - name: windowswebserver
        image: mcr.microsoft.com/windows/servercore:ltsc2019
        imagePullPolicy: IfNotPresent
        command:
        - powershell.exe
        - -command
        - $listener = New-Object System.Net.HttpListener; $listener.Prefixes.Add('http://*:80/'); $listener.Start();Write-Host('Listening at http://*:80/'); while ($listener.IsListening) { $context = $listener.GetContext(); $response = $context.Response; $content='<html><body><H1>Windows Container Web Server</H1></body></html>'; $buffer = [System.Text.Encoding]::UTF8.GetBytes($content); $response.ContentLength64 = $buffer.Length; $response.OutputStream.Write($buffer, 0, $buffer.Length); $response.Close(); };
        volumeMounts:
        - mountPath: "C:\\Data"
          name: volume
      volumes:
      - name: volume
        persistentVolumeClaim:
          claimName: azure-file-win
      nodeSelector:
        beta.kubernetes.io/os: windows
```
Then run `curl -s -k -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"https://<NODE_IP>:10250/stats/summary` where <NODE_IP> is the IP of the Windows node the pod is running on and view the output for the pod.

### Kubernetes version

1.21.7

### Cloud provider

Azure

### OS version

Caption: Microsoft Windows Server 2019 Datacenter
Version: 10.0.17763
BuildNumber:17763
OSArchitecture: 64-bit

### Container runtime (CRI) and version (if applicable)

Docker

### Related plugins (CNI, CSI, ...) and versions (if applicable)

AzureFile CSI Driver version: 1.17.0

**Comment by andyzhangx** [comment]
I have checked this issue, this is actually related to this issue: https://github.com/kubernetes-csi/csi-proxy/issues/208
There is no such good way to get usage of smb mount on Windows node per my knowledge.
/sig windows

**Comment by SergeyKanzhelev** [comment]
/remove-sig node

since it looks to be a known windows issue

**Comment by dcantah** [comment]
Asked some SMB folks. In the meantime I've been lookin around and https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getdiskfreespaceexa may be an avenue. Believe it works on smb shares,I can report back tomorrow.

**Comment by dcantah** [comment]
GetDiskFreeSpaceEx seems to work from testing on an az file share.

```go
package main

import (
	"fmt"
	"log"
	"os"

	"golang.org/x/sys/windows"
)

func main() {
	var (
		total     uint64
		totalFree uint64
	)
	dirName := windows.StringToUTF16Ptr(os.Args[1])
	if err := windows.GetDiskFreeSpaceEx(dirName, nil, &total, &totalFree); err != nil {
		log.Fatal(err)
	}
	toGib := func(bytes uint64) int {
		return int(bytes >> 30)
	}
	fmt.Printf("Total: %d\nTotal Free: %d\n", toGib(total), toGib(totalFree))
}
```

The above reports:
Total: 5120
Total Free: 5100

with the share I was testing with having a 5TiB quota and a single 20 GiB file inside. I'll wait for the SMB folks to tell me if there's any gotchas, but if not I can push a change to add this to the csi-proxy

**Comment by andyzhangx** [comment]
@dcantah what's the `os.Args[1]`? is it `\\MyServer\MyShare\` or mounted directory name?

**Comment by dcantah** [comment]
@andyzhangx \\myserver\myshare\

**Comment by andyzhangx** [comment]
@dcantah The k8s file driver would create a symlink to the azure file, like following:
```
C:\var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-f9b134da-b69d-46bf-9aaf-3787476ed2b1>dir
 Volume in drive C has no label.
 Volume Serial Number is AAFE-3BA6

 Directory of C:\var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-f9b134da-b69d-46bf-9aaf-3787476ed2b1

06/05/2022  12:37 PM    <DIR>          .
06/05/2022  12:37 PM    <DIR>          ..
06/05/2022  12:37 PM    <SYMLINKD>     globalmount [\\xxx.file.core.windows.net\pvc-f9b134da-b69d-46bf-9aaf-3787476ed2b1\]
06/05/2022  12:37 PM               149 vol_data.json
               1 File(s)            149 bytes
               3 Dir(s)  108,333,318,144 bytes free
```
Would `GetDiskFreeSpaceEx` work for both symlink and UNC path?

**Comment by dcantah** [comment]
@andyzhangx Looks like it:
```
PS C:\Users\danny\go\src\github.com\dcantah\freespace> .\freespace.exe C:\smbtest\                 
Total: 5120
Total Free: 5100
```

**Comment by andyzhangx** [comment]
@dcantah great, current `(VolumeAPI) GetVolumeStats(volumeID string)` accepts a volumeID which is only for disk, I think you could add a new API `(VolumeAPI) GetVolumeStatsFromPath(path string)` accepts a mounted dir, that should work for both disk and smb mounts in csi-proxy, wdyt?

https://github.com/kubernetes-csi/csi-proxy/blob/48ba409159027564c50926138dd951d106b53077/pkg/os/volume/api.go#L201-L225

**Comment by marosset** [comment]
/remove-lifecycle stale
@dcantah friendly ping on the above question :)

**Comment by lualvare** [comment]
Hello @dcantah, hope you are doing well.

I am following up into this issue to understand if you were able to get this to work. If yes, could you please share the details here?

Thank you so much.

**Comment by andyzhangx** [comment]
> Hello @dcantah, hope you are doing well.
> 
> I am following up into this issue to understand if you were able to get this to work. If yes, could you please share the details here?
> 
> Thank you so much.

@lualvare this requires csi-proxy change and new csi-proxy version upgrade. And we are in the process of removing csi-proxy and use native call from k8s 1.23, after that csi-proxy change, it's easier to fix this issue in csi driver.

**Comment by marosset** [comment]
@kiashok FYI

**Comment by lualvare** [comment]
@andyzhangx  thank you so very much for the information provided here, I want to know if this is already in the roadmap and see if there is any  ETA for this fix?

Thank you so much.

**Comment by andyzhangx** [comment]
> @andyzhangx thank you so very much for the information provided here, I want to know if this is already in the roadmap and see if there is any ETA for this fix?
> 
> Thank you so much.

@lualvare it mainly depends on when csi-proxy removal would be completed: https://github.com/kubernetes-csi/csi-proxy/issues/217, since we don't publish any new version of csi-proxy now, after that work is done, such kind of change is easier. There is no clear ETA yet.

**Comment by andyzhangx** [comment]
FYI. this issue would be fixed on AKS 1.27 with host process deployment directly, here is an example PR how to fix this issue: https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/1337
